### PR TITLE
reimplement FileListView with BColumnListView

### PR DESCRIPTION
**Pull request by hannahyp**
This is what I have so far. It crashes most of the time when selecting files, but I can't figure out why.
pecorename
crash
debug

### CODE REVIEW

**Comment by hannahyp** [comment]
A debug report file:
[PecoRename-972-debug.report.txt](https://github.com/HaikuArchives/PecoRename/files/99029/PecoRename-972-debug.report.txt)

**Comment by hannahyp** [comment]
Thanks, Adrien! It's fixed now.

pecorename

File: Main/FileListItem.h [hunk 1]
```diff
@@ -31,6 +32,8 @@ friend	void PecoApp::CreateScript(BMessage *msg);
 	void			SetNewName( BString	myNewName );
 	void 			SetName( BString name );
 	bool			CompareWith(FileListItem *CompareItem);
+	void			SetRow(FileListRow* row) { fRow = row; }
+	FileListRow*	GetRow() const { return fRow; }
```
Comment (pulkomandy) [inline Main/FileListItem.h]: There seem to be a problem with the indentation here?

File: Main/FileListView.cpp [hunk 2]
```diff
@@ -22,16 +23,27 @@
...
 void FileListView::MouseDown(BPoint where) {
-	BListView::MouseDown(where);
+	BColumnListView::MouseDown(where);
 //	DeselectAll();
 };
 
-bool FileListView::InitiateDrag(BPoint where, int32 itemIndex, bool initialySelected) {
+bool FileListView::InitiateDrag(BPoint where, bool initialySelected) {
+	int32	itemIndex = IndexOf(RowAt(where));
 	FileListItem	*Item = (FileListItem *)ItemAt(itemIndex);
```
Comment (pulkomandy) [inline Main/FileListView.cpp]: This is superfluous, you can get the item first (using RowAt), then the index (using IndexOf), and avoid the extra ItemAt call.

File: Main/FileListView.cpp [hunk 3]
```diff
@@ -74,7 +86,7 @@ bool FileListView::InitiateDrag(BPoint where, int32 itemIndex, bool initialySele
 };
 
 void FileListView::KeyDown(const char *bytes, int32 numBytes) {
-	int32	selectedItem = CurrentSelection();
+	int32	selectedItem = IndexOf(CurrentSelection());
 	FileListItem	*Item = (FileListItem *)ItemAt(selectedItem);
```
Comment (pulkomandy) [inline Main/FileListView.cpp]: Same here, no need for ItemAt.

File: Main/FileListView.h [hunk 4]
```diff
@@ -8,12 +8,35 @@
...
+
+class FileListView : public BColumnListView {
 public:
 					FileListView(BRect frame);
 	virtual void	MouseDown(BPoint where);
 	virtual	void	KeyDown(const char *bytes, int32 numBytes);
-	virtual bool	InitiateDrag(BPoint pt, int32 itemIndex, bool initialySelected);
+	virtual bool	InitiateDrag(BPoint pt, bool initialySelected);
+	FileListItem*	ItemAt(int32 index) const;
+			void	Select(int32 index);
```
Comment (pulkomandy) [inline Main/FileListView.h]: Indentation problems here again

File: Main/FileListView.h [hunk 5]
```diff
@@ -8,12 +8,35 @@
...
+			void	InvalidateItem(int32 index);
+			void	AddItem(FileListItem* item);
+			void	AddList(BList* list);
+			void	RemoveItem(FileListItem* item);
+			void	MakeEmpty() { Clear(); }
+			int32	CountItems() const { return CountRows(); }
```
Comment (pulkomandy) [inline Main/FileListView.h]: I think for such simple methods (at least MakeEmpty and CountItems), it would be better to replace the code in all place where it is called?

File: Main/FileView.cpp [hunk 6]
```diff
@@ -26,6 +26,7 @@ FileView::FileView(BRect frame) : BView( frame, "fileView", B_FOLLOW_ALL, B_WILL
 };
 
 void FileView::AttachedToWindow() {
+/*
```
Comment (pulkomandy) [inline Main/FileView.cpp]: Why was this removed?
It would be preferred to use #if 0 to disable code, and add a comment explaining why it is disabled.
Or if the code is now completely unneeded, remove it completely.

File: Main/FileListItem.h (hunk 1 above)
Comment (hannahyp) [inline Main/FileListItem.h]: No. I followed the indentation of the original code (tab stop = 4). Here it is displayed with tab stop = 8.

File: Main/FileListView.cpp (hunk 2 above)
Comment (hannahyp) [inline Main/FileListView.cpp]: BColumnListView::RowAt returns a BRow, not BListItem, and BColumnListView::IndexOf only works on a BRow but not BListItem.

The alternative way (which I had used before) would be:

FileListRow* row = (FileListRow*) RowAt(where);
FileListItem* Item = row->Item();
int32 itemIndex = IndexOf(row);

which I think is a little faster but with one extra line of code. I have switched it back to this earlier version.

File: Main/FileListView.cpp (hunk 3 above)
Comment (hannahyp) [inline Main/FileListView.cpp]: I have also switched it back to the earlier version I had.

File: Main/FileListView.h (hunk 4 above)
Comment (hannahyp) [inline Main/FileListView.h]: Same here. (No indentation problems.)

File: Main/FileListView.h (hunk 5 above)
Comment (hannahyp) [inline Main/FileListView.h]: I updated the code.

File: Main/FileView.cpp (hunk 6 above)
Comment (hannahyp) [inline Main/FileView.cpp]: Done. Thanks for your invaluable help!

**Comment by pulkomandy** [comment]
Is there a reason to not make FileListItem extend BRow directly (instead of using the FileListRow wrapper class)? I think this could make the code a little simpler.

**Comment by hannahyp** [comment]
I think it would impact many more files, and many more places in those files (like PecoApp.cpp). Anywhere a FileListItem or BListItem  is used, it must be changed to a BRow plus converting BListItem methods to corresponding BRow counterparts if they exist or reimplementing them if they don't.
